uvm.go

```diff
@@ -0,0 +1,80 @@
+package hcsshim
+
+import (
+	"time"
+
+	"github.com/Microsoft/hcsshim/internal/schemaversion"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+// CreateUtilityVMOptions are the set of options which are passed to NewUtilityVM() to
+// create a utility vm.
+type CreateUtilityVMOptions struct {
+	Id                      string                  // Identifier for the uvm. Defaults to generated GUID.
+	Owner                   string                  // Specifies the owner. Defaults to executable name.
+	OperatingSystem         string                  // "windows" or "linux". // TODO: We can probably deduce this in a number of ways.
+	Resources               *specs.WindowsResources // Optional resources for the utility VM. Supports Memory.limit and CPU.Count only currently. // TODO consider extending?
+	AdditionHCSDocumentJSON string                  // Optional additional JSON to merge into the HCS document prior
+
+	// WCOW specific parameters
+	LayerFolders []string // Set of folders for base layers and sandbox. Ordered from top most read-only through base read-only layer, followed by sandbox
+
+	// LCOW specific parameters
+	KirdPath               string                       // Folder in which kernel and initrd reside. Defaults to \Program Files\Linux Containers
+	KernelFile             string                       // Filename under KirdPath for the kernel. Defaults to bootx64.efi
+	InitrdFile             string                       // Filename under KirdPath for the initrd image. Defaults to initrd.img
+	KernelBootOptions      string                       // Additional boot options for the kernel
+	KernelDebugMode        bool                         // Configures the kernel in debug mode using sane defaults
+	KernelDebugComPortPipe string                       // If kernel is in debug mode, can override the pipe here.
+	SchemaVersion          *schemaversion.SchemaVersion // For a v1 service VM (back-compatibility). Assumes v2 unless explicitly provided
```

Comment:
If we need to support LCOW on pre-RS5 using the v1 schema, I'll need this. Here a "utility VM" is what is used as a "service VM" in the graphdriver today.

Reply:
I think if we want a Krypton then we should use a raw hcs.System and not overload this utility VM object.

---
In reply to: [189366176](https://github.com/Microsoft/hcsshim/pull/192#discussion_r189366176) [](ancestors = 189366176)